### ISSUE TITLE
`ModelBridge.evaluate_acquisition_function` returns a length-one list regardless of how many observation features were specified

### ISSUE THREAD

**Issue by lena-kashtelyan** [facebook/Ax#778]
Originally reported here: https://github.com/facebook/Ax/issues/771#issuecomment-1013311624

**Comment by sgbaird** [comment]
Could this be the issue?
```python
evaluate (c:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\models\torch\botorch_modular\acquisition.py:316)
evaluate_acquisition_function (c:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\models\torch\botorch_modular\model.py:341)
_model_evaluate_acquisition_function (c:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\torch.py:134)
_evaluate_acquisition_function (c:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\array.py:494)
evaluate_acquisition_function (c:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\base.py:866)
```
The `evaluate` docstring looks like this:
```python
    def evaluate(self, X: Tensor) -> Tensor:
        """Evaluate the acquisition function on the candidate set `X`.

        Args:
            X: A `batch_shape x q x d`-dim Tensor of t-batches with `q` `d`-dim design
                points each.

        Returns:
            A `batch_shape'`-dim Tensor of acquisition values at the given
            design points `X`, where `batch_shape'` is the broadcasted batch shape of
            model and input `X`.
        """
```
However, my `X` is 2D, not 3D. `X.shape` == `torch.Size([32, 4])`

**Comment by sgbaird** [comment]
https://github.com/facebook/Ax/blob/5e73786f466a965772f2540fb74cbf1092967f23/ax/models/torch/botorch_modular/acquisition.py#L302-L316

Changing Line 316 to:
```python
        return self.acqf(X=torch.unsqueeze(X, 1))
```
seems to produce the desired shape of `acqf_values`. See #779

**Comment by sgbaird** [comment]
After implementing the fix from #779, I get the following OOM error with 32768 candidates:
```python
[enforce fail at ..\c10\core\CPUAllocator.cpp:76] data. DefaultCPUAllocator: not enough memory: you tried to allocate 13019119616 bytes.
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\lazy\non_lazy_tensor.py", line 44, in _matmul
    return torch.matmul(self.tensor, rhs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\functions\_matmul.py", line 21, in forward
    res = lazy_tsr._matmul(rhs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\lazy\lazy_tensor.py", line 1389, in matmul
    return Matmul.apply(self.representation_tree(), other, *self.representation())
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\lazy\triangular_lazy_tensor.py", line 79, in _matmul
    return self._tensor.matmul(rhs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\functions\_matmul.py", line 21, in forward
    res = lazy_tsr._matmul(rhs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\lazy\lazy_tensor.py", line 1389, in matmul
    return Matmul.apply(self.representation_tree(), other, *self.representation())
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\gpytorch\distributions\multivariate_normal.py", line 212, in rsample
    res = covar_root.matmul(base_samples) + self.loc.unsqueeze(-1)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\botorch\posteriors\gpytorch.py", line 95, in rsample
    samples = self.mvn.rsample(
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\botorch\sampling\samplers.py", line 93, in forward
    samples = posterior.rsample(
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\botorch\acquisition\monte_carlo.py", line 245, in forward
    samples = self.sampler(posterior)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\botorch\utils\transforms.py", line 229, in decorated
    output = method(acqf, X, *args, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\botorch\utils\transforms.py", line 270, in decorated
    return method(cls, X, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\models\torch\botorch_modular\acquisition.py", line 316, in evaluate
    return self.acqf(X=torch.unsqueeze(X, 1))
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\models\torch\botorch_modular\model.py", line 341, in evaluate_acquisition_function
    return acqf.evaluate(X=X)
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\torch.py", line 134, in _model_evaluate_acquisition_function
    evals = not_none(self.model).evaluate_acquisition_function(
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\array.py", line 494, in _evaluate_acquisition_function
    return self._model_evaluate_acquisition_function(
  File "C:\Users\sterg\miniconda3\envs\dental-composite\Lib\site-packages\ax\modelbridge\base.py", line 866, in evaluate_acquisition_function
    return self._evaluate_acquisition_function(
  File "C:\Users\sterg\Documents\GitHub\sparks-issa\Ultradent-BO\dental_composite_opt\service_predefined.py", line 218, in <module>
    acqf_values = model_bridge.evaluate_acquisition_function(
```

**Comment by Balandat** [comment]
Correct - if you pass a tensor of shape `q x d`, this will be considered a batch of `q` candidates and the acquisition function will compute the *joint* acquisition value of that batch. So if you did use `q=32768` that would mean computing and sampling from a `32768`-dimensional multivariate normal. 

The fix you have in #779 makes sense in your case if you are only interested in the case `q=1`, but we also should allow evaluating the joint value of a batch. One obvious thing we could do is change the API to allow passing in lists of lists of observation features and then convert those to a tensor of the appropriate shape. Another option would be to add sth like a `is_batch` argument, but I like the lists of lists setup better - if it's just a single flat list we can interpret as a batch of individual points (which is probably what most people want), an if someone passes in a nested list we can do the right thing with `q`-batches. @lena-kashtelyan, @bletham any other thoughts on a good interface for this? 

The reason you're seeing OOMs with your change in #779 is that for `NEI` we're evaluating and sampling from `32768` posterior distributions over `[X_train, X_i]` for `i=1, ..., 32768` (maybe a subset of `X_train` if we're pruning, but still many points), so there just needs to be a lot of memory allocated for all those matrices and the involved computations. Chunking things into smaller batches is the appropriate workaround for this.

**Comment by sgbaird** [comment]
Thank you! This provides a lot of context for the problem. Interestingly, the list of lists idea was actually one of the early things I tried, so I think it would be intuitive. What about the output? If you go with the list of lists input, should it still be a flat list or should it also be a list of lists?

**Comment by Balandat** [comment]
It should return one acquisition value for each list in the list. Each inner list will be consider as one `q`-batch with an associated joint acquisition value.

**Comment by sgbaird** [comment]
Ok, gotcha. After reading [the documentation](https://botorch.org/docs/batching#batch-acquisition-functions) a bit closer, I'm seeing that behavior described.

For a single set of `32768` design candidates (`candidates : List[ObservationFeatures]`), how would I get 10 next suggested experiments instead of just one? I think I understand that, memory constraints aside, the input tensor would be of `shape([32768, 10, d])` in the `b × q × d` sense from the [batching docs](https://botorch.org/docs/batching). However, I'm not exactly sure what this looks like in practice. Is it simply repeating the list of candidates 10 times and swapping the first and second dimensions (`batch_candidates : List[List[ObservationFeatures]] = np.swap_axes(10 * [candidates], 0, 1)`) to get `shape([32768, 10, d])`? I think this is where my understanding breaks down, but I get the sense it might not be possible to do chunking (my workaround) with joint optimization since I'd only pass in a small subset of the data to evaluate_acquisition_function` at a time.

It seems like the motivation of joint candidate optimization is to find the set of `q` candidate suggestions (e.g. 10 candidates) simultaneously, and sequential optimization uses successive conditioning to find the same. Does successive conditioning mean performing an inner loop of "fantasy" adaptive design where the model predictions are sequentially added back into the model as if they were measured values? If so, then this seems like something I can do even with my [chunking/outer-for-loop](https://github.com/facebook/Ax/issues/778#issuecomment-1013529560) approach to avoid OOM errors.

**Comment by Balandat** [comment]
You're getting into the subtle points here, that's great.

You are correct that if there are `K` feasible parameterizations, and you want to pick the set with `q` elements that have the highest joint acquisition value you have to consider `K` choose `q` sets. Enumerating these is usually not feasible (unless `K` and `q` are small). This is usually not feasible, so a sequential greedy / successive conditioning approach is typically used that picks the highest-valued point and conditions on previously selected points. Due to the submodularity of the acquisition function, you can actually show that the regret of this procedure relative to the optimal one is limited (to 1/e) as shown by @j-wilson in https://arxiv.org/abs/1805.10196. 

This can be done in (at least) two ways: One is the "fantasy" model that you described. What we typically do instead is use a MC approach for this described in Appendix F.2 of https://arxiv.org/pdf/1910.06403.pdf. 

And yes, you can do this using your chunking-based approach, by passing previously selected points in via `pending_observations` to `evaluate_acquisition_function`.

**Comment by Balandat** [comment]
@sgbaird I believe this should largely be resolved once #779 goes in, correct?

**Comment by sgbaird** [comment]
@Balandat that's correct

**Comment by saitcakmak** [comment]
Closed by #779

**Comment by saitcakmak** [comment]
I guess the proper thing to do is to mark it "fix-ready" until the next release :)

**Comment by lena-kashtelyan** [comment]
The fix for this is now on latest stable: 0.2.5.1 : )
